### ISSUE TITLE
Is there a way to limit the API request rate?

### ISSUE THREAD

**Issue by ShahanaFarooqui** [Ride-The-Lightning/RTL#1067]
### Discussed in https://github.com/Ride-The-Lightning/RTL/discussions/1066

<div type='discussions-op-text'>

<sup>Originally posted by **DerEwige** August 16, 2022</sup>
My RTL freezes (overloads) my eclair node when it tries to load the invoices.
(When I go to lightning -> transactions)

The node CPU goes up to 100% (4 cores) and I get time out errors in the logs for http://localhost:8080/listinvoices

I asume that RTL has problem handling over a million of invoices.
Is there an option in the config to tell RTL to load the data slower?
</div>

**Comment by DerEwige** [comment]
Some additional info regarding this problem.

The issue happens because I have several hundred thousend invoices in my database.
RTL tries to fetch them all at once. This results in a timout of the API call and then RTL tries again.

Solution: use the from and to fields of the listinvoice API point and if it gives a timeout reduce the time to get smaller junks

**Comment by ShahanaFarooqui** [comment]
@DerEwige Thank you for the issue.

After analysing it more closely, we believe that it will be better handled if pagination option is added for the API. I opened two issues requesting pagination on `listinvoices` and `audit` APIs as below: 
https://github.com/ACINQ/eclair/issues/2450
https://github.com/ACINQ/eclair/issues/2451

I also opened another issue to delete any unpaid expired invoice in order to remove unnecessary records from the node:  
https://github.com/ACINQ/eclair/issues/2449

I am putting this issue on hold until at least [2450](https://github.com/ACINQ/eclair/issues/2450) is addressed.

**Comment by ShahanaFarooqui** [comment]
@DerEwige Eclair already provides an option to delete all unpaid expired invoices. Please feel free to configure if you have a lot of expired invoices. 

> https://github.com/ACINQ/eclair/pull/2210/files
```
Expired incoming invoices that are unpaid will be searched for and purged from the database when Eclair starts up. Thereafter searches for expired unpaid invoices to purge will run once every 24 hours. You can disable this feature, or change the search interval with two new settings:

* `eclair.purge-expired-invoices.enabled = true
* `eclair.purge-expired-invoices.interval = 24 hours`
```

We will improve both API request as soon as pagination is available.

**Comment by DerEwige** [comment]
Thanks @ShahanaFarooqui for sharing
Once the new release drops. I will defenitly use this

**Comment by ShahanaFarooqui** [comment]
Unblocked with PR https://github.com/ACINQ/eclair/pull/2474

**Comment by mviswanathsai** [comment]
Hey! I am looking to contribute, can someone guide me on where I can start?

**Comment by ShahanaFarooqui** [comment]
Hey, you can start by following https://github.com/Ride-The-Lightning/RTL/blob/master/.github/docs/Contributing.md and start a new thread in discussions board if more information is needed after that.

**Comment by mviswanathsai** [comment]
I have cloned the repository and tried running the server, all is fine. But
I am just lost as to where to start since I have not tried meddling with a
code base as big. I would be very helpful if you could guide me around what
kind of beginner issues I can tackle first to get a feel for the system.
That way I can overcome my inertia and contribute, since I plan on being
regular.

Thanks for reaching out!

**Comment by ShahanaFarooqui** [comment]
Pagination cannot be implemented without API responding with total number of records which is missing currently.
I will wait for the issue [2855](https://github.com/ACINQ/eclair/issues/2855) to add this in the response before implementing it in the UI.

**Comment by ShahanaFarooqui** [comment]
Updated code to enable pagination in future (whenever APIs are fixed) with PR #1393.

**Comment by saubyk** [comment]
Status update (2026-08-18) — the upstream picture has changed, so re-triaging:

- **The Eclair blockers are resolved, one of them negatively.** `count`/`skip` pagination landed on `listinvoices` and `audit` in 2022 (ACINQ/eclair#2450, #2451). The remaining ask — a total-record count in the responses so the paginator can show page N of M (ACINQ/eclair#2855) — was **closed as inactive on 2026-01-02**, with the Eclair maintainer explicitly questioning the need for a total ("UX has been going towards infinite scroll"). So there is nothing left to wait for upstream.
- **RTL is half-way there.** #1393 plumbed `count`/`skip` end to end, and the paginator's page-change handler already dispatches `count: pageSize, skip: page × size`. But the *initial* load in `lightning-invoices.component.ts` still fetches everything (`count: 1000000, skip: 0`), with the paginated first-page dispatch commented out "until the paginator API is fixed" — so the original symptom (a full-table `listinvoices` on opening Transactions) is still the behaviour today.

Since the "uncomment after the API is fixed" plan is dead, the fix now needs a UX decision rather than more waiting: paginate **without a known total** (server-side paging where "next" stays enabled while a full page comes back — the same shape LND's `index_offset` paging has), or a bounded first load with "load more". Either is implementable now. The audit/payments path is bounded by date already, so invoices is the acute one.

Removing `blocked` and taking this off the 0.15.11 milestone — it's a medium frontend change for a later release, not something to hold 0.15.11 for.

**Comment by saubyk** [comment]
Fixed by #1690, merged into `Release-0.15.12`: the backend now derives the invoice total with O(log n) single-row probes and serves one newest-first page (default 10, capped at 100), so the full-table `listinvoices` and the per-invoice `/getreceivedinfo` fan-out are gone. Ships with v0.15.12.
